### ISSUE TITLE
Disallow plain toString calls as part of no-implicit-coercion

### ISSUE THREAD

**Issue by remcohaszing** [eslint/eslint#13811]
**What rule do you want to change?**

`no-implicit-coercion`

**Does this change cause the rule to produce more or fewer warnings?**

More

**How will the change be implemented? (New option, new default behavior, etc.)?**

The rule will report usage of `.toString()`.

A new option `toString` will be added. In a major release this option will be removed and behaviour will be merged with the existing `string` option.

The following edge cases will **not** be reported:

```js
foo?.toString()
foo.toString('ascii')
foo.toString
```

This may change behaviour of the code. For example `null.toString()` will be converted to `String(null)`, which will no longer throw.

**Please provide some example code that this change will affect:**

```js
foo.toString()
```

**What does the rule currently do for this code?**

It will report the `foo.toString()` call expression.

**What will the rule do after it's changed?**

The code will be replaced with the following:

```js
String(foo)
```

**Are you willing to submit a pull request to implement this change?**

Yes

**Comment by mdjermanovic** [comment]
Thanks for the enhancement proposal!

I'm not sure if this belongs to `no-implicit-coercion`, since it's an explicit method call.

Also, in addition to the `null` example, there are other cases where `.toString()` doesn't have to be the same as `String()`:

```js
var foo = { toString() { return 5; } };

foo.toString(); // 5
String(foo); // "5"
```

```js
var foo = { toString() { return {}; }, valueOf() { return 5 } };

foo.toString(); // {}
String(foo); // "5"
```

```js
var foo = { toString() { return {}; } };

foo.toString(); // {}
String(foo); // throws TypeError
```

```js
var foo = { __proto__: null, valueOf() { return 5; } };

foo.toString(); // throws TypeError
String(foo); // "5"
```

These examples are not common, but still it doesn't look safe to auto-fix `foo.toString()` to `String(foo)`.

(albeit, the rule already fixes `"" + foo` to `String(foo)` which isn't the same in some edge cases)

Without auto-fixing, `no-restricted-syntax` could do the work.

Given that `foo.toString()` isn't quite _implicit_, and probably isn't fixable, I'm not in favor of adding this check to `no-implicit-coercion`, but wouldn't be opposed if other team members have a different opinion.

**Comment by nzakas** [comment]
I agree with @mdjermanovic, `toString()` is an explicit call and it doesn't make sense to add to this rule.

**Comment by remcohaszing** [comment]
I agree `toString()` is pretty explicit. For me it’s more about consistency.

Currently I’m using `no-restricted-syntax` as @mdjermanovic mentioned, but unfortunately this isn’t auto fixable.

**Comment by nzakas** [comment]
@remcohaszing this sounds like a situation where creating a custom rule is the best approach.

**Comment by nzakas** [comment]
Closing, as there is no change to be made here.
